Remove leading and trailing spaces in title

When adding new images through the app it happens that the caption ends with a space. This is then used for title/filename and produces ugly links. Please run a `.trim()` before uploading.

I'm helping someone learn the app, and they're hitting this bug quite often because they inadvertently add a space at the end of the caption. This means they hit the titleblacklist-custom-space issue (because the filename ends up as `North Cottesloe beach .jpg`):

```
2023-01-07 09:32:29,004     [ERROR] [file-logging-thread-1] [UploadClient] : Error in uploading file from stash
org.wikipedia.dataclient.mwapi.MwException: ⧼titleblacklist-custom-space⧽

2023-01-07 09:32:29,005     [ERROR] [file-logging-thread-1] [UploadWorker] : java.lang.NullPointerException: The supplied value is null
	at io.reactivex.internal.operators.observable.ObservableOnErrorReturn$OnErrorReturnObserver.onError(ObservableOnErrorReturn.java:80)
	at io.reactivex.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
	at io.reactivex.internal.observers.BasicFuseableObserver.fail(BasicFuseableObserver.java:110)
	at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:59)
	at retrofit2.adapter.rxjava2.BodyObservable$BodyObserver.onNext(BodyObservable.java:51)
	at retrofit2.adapter.rxjava2.BodyObservable$BodyObserver.onNext(BodyObservable.java:37)
	at retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:44)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:34)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableOnErrorReturn.subscribeActual(ObservableOnErrorReturn.java:31)
	at io.reactivex.Observable.subscribe(Observable.java:12267)
	at io.reactivex.internal.operators.observable.ObservableSingleMaybe.subscribeActual(ObservableSingleMaybe.java:31)
	at io.reactivex.Maybe.subscribe(Maybe.java:4290)
	at io.reactivex.Maybe.blockingGet(Maybe.java:2320)
	at io.reactivex.Observable.blockingSingle(Observable.java:5381)
	at fr.free.nrw.commons.upload.worker.UploadWorker.uploadContribution(UploadWorker.kt:300)
	at fr.free.nrw.commons.upload.worker.UploadWorker.access$uploadContribution(UploadWorker.kt:48)
	at fr.free.nrw.commons.upload.worker.UploadWorker$doWork$2$invokeSuspend$$inlined$map$1$2.emit(Collect.kt:148)
	at kotlinx.coroutines.flow.FlowKt__BuildersKt$asFlow$$inlined$unsafeFlow$3.collect(SafeCollector.common.kt:115)
	at fr.free.nrw.commons.upload.worker.UploadWorker$doWork$2$invokeSuspend$$inlined$map$1.collect(SafeCollector.common.kt:114)
	at kotlinx.coroutines.flow.FlowKt__CollectKt.collect(Collect.kt:30)
	at kotlinx.coroutines.flow.FlowKt.collect(Unknown Source:1)
	at fr.free.nrw.commons.upload.worker.UploadWorker$doWork$2.invokeSuspend(UploadWorker.kt:218)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.Exception: titleblacklist-custom-space
	at fr.free.nrw.commons.upload.UploadClient.lambda$uploadFileFromStash$3$UploadClient(UploadClient.java:227)
	at fr.free.nrw.commons.upload.-$$Lambda$UploadClient$-xVSLroq3t059tyyfYuDUQukbVk.apply(Unknown Source:4)
	at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:57)
	... 28 more

2023-01-07 09:32:29,006     [ERROR] [file-logging-thread-1] [UploadWorker] : Upload from stash failed for contribution : North Cottesloe beach .jpg
```

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java
@@ -1,7 +1,6 @@
 package fr.free.nrw.commons.upload;
 
 import android.app.Dialog;
-import android.content.Intent;
 import android.text.Editable;
 import android.text.InputFilter;
 import android.text.TextUtils;
@@ -23,7 +22,6 @@ import butterknife.BindView;
 import butterknife.ButterKnife;
 import com.google.android.material.textfield.TextInputLayout;
 import fr.free.nrw.commons.R;
-import fr.free.nrw.commons.contributions.MainActivity;
 import fr.free.nrw.commons.recentlanguages.Language;
 import fr.free.nrw.commons.recentlanguages.RecentLanguagesAdapter;
 import fr.free.nrw.commons.recentlanguages.RecentLanguagesDao;
@@ -32,10 +30,8 @@ import fr.free.nrw.commons.utils.AbstractTextWatcher;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
-import java.util.Locale;
 import java.util.Objects;
 import java.util.regex.Pattern;
-import javax.inject.Inject;
 import timber.log.Timber;
 
 public class UploadMediaDetailAdapter extends RecyclerView.Adapter<UploadMediaDetailAdapter.ViewHolder> {
@@ -198,7 +194,8 @@ public class UploadMediaDetailAdapter extends RecyclerView.Adapter<UploadMediaDe
 
             removeButton.setOnClickListener(v -> removeDescription(uploadMediaDetail, position));
             captionListener = new AbstractTextWatcher(
-                captionText -> uploadMediaDetails.get(position).setCaptionText(convertIdeographicSpaceToLatinSpace(removeTrailingWhitespace(captionText))));
+                captionText -> uploadMediaDetails.get(position).setCaptionText(convertIdeographicSpaceToLatinSpace(
+                    removeLeadingAndTrailingWhitespace(captionText))));
             descriptionListener = new AbstractTextWatcher(
                 descriptionText -> uploadMediaDetails.get(position).setDescriptionText(descriptionText));
             captionItemEditText.addTextChangedListener(captionListener);
@@ -434,15 +431,33 @@ public class UploadMediaDetailAdapter extends RecyclerView.Adapter<UploadMediaDe
         }
 
         /**
-         * Removes any trailing whitespace from the source text.
+         * Removes any leading and trailing whitespace from the source text.
          * @param source input string
-         * @return a string without trailing whitespace
+         * @return a string without leading and trailing whitespace
          */
-        public String removeTrailingWhitespace(String source) {
-            while (checkTrailingWhitespace(source)) {
-                source = source.substring(0, source.length() - 1);
+        public String removeLeadingAndTrailingWhitespace(String source) {
+            int firstNonWhitespaceIndex = 0;
+            while (firstNonWhitespaceIndex < source.length()) {
+                if (Character.isWhitespace(source.charAt(firstNonWhitespaceIndex))) {
+                    firstNonWhitespaceIndex++;
+                } else {
+                    break;
+                }
             }
-            return source;
+            if (firstNonWhitespaceIndex == source.length()) {
+                return "";
+            }
+
+            int lastNonWhitespaceIndex = source.length() - 1;
+            while (lastNonWhitespaceIndex > firstNonWhitespaceIndex) {
+                if (Character.isWhitespace(source.charAt(lastNonWhitespaceIndex))) {
+                    lastNonWhitespaceIndex--;
+                } else {
+                    break;
+                }
+            }
+
+            return source.substring(firstNonWhitespaceIndex, lastNonWhitespaceIndex + 1);
         }
 
         /**

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java
@@ -423,6 +423,8 @@ public class UploadMediaDetailAdapter extends RecyclerView.Adapter<UploadMediaDe
          * @return a string without leading and trailing whitespace
          */
         public String removeLeadingAndTrailingWhitespace(String source) {
+            // This method can be replaced with the inbuilt String::strip when updated to JDK 11.
+            // Note that String::trim does not adequately remove all whitespace chars.
             int firstNonWhitespaceIndex = 0;
             while (firstNonWhitespaceIndex < source.length()) {
                 if (Character.isWhitespace(source.charAt(firstNonWhitespaceIndex))) {

--- a/app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java
+++ b/app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java
@@ -418,19 +418,6 @@ public class UploadMediaDetailAdapter extends RecyclerView.Adapter<UploadMediaDe
         }
 
         /**
-         * Checks if the source string contains trailing whitespace
-         * @param source input string
-         * @return true if contains trailing whitespace and false otherwise
-         */
-        public Boolean checkTrailingWhitespace(String source) {
-            int len = source.length();
-            if (len == 0) {
-                return false;
-            }
-            return Character.isWhitespace(source.charAt(len - 1));
-        }
-
-        /**
          * Removes any leading and trailing whitespace from the source text.
          * @param source input string
          * @return a string without leading and trailing whitespace

--- a/app/src/test/kotlin/fr/free/nrw/commons/upload/UploadMediaDetailAdapterUnitTest.kt
+++ b/app/src/test/kotlin/fr/free/nrw/commons/upload/UploadMediaDetailAdapterUnitTest.kt
@@ -248,39 +248,39 @@ class UploadMediaDetailAdapterUnitTest {
     }
 
     @Test
-    fun testRemoveTrailingWhitespace() {
+    fun testRemoveLeadingAndTrailingWhitespace() {
         // empty space
-        val test1 = "test  "
+        val test1 = "  test  "
         val expected1 = "test"
-        Assert.assertTrue(viewHolder.checkTrailingWhitespace(test1));
-        Assert.assertEquals(expected1, viewHolder.removeTrailingWhitespace(test1))
+        Assert.assertEquals(expected1, viewHolder.removeLeadingAndTrailingWhitespace(test1))
 
-        val test2 = "test test "
+        val test2 = "  test test "
         val expected2 = "test test"
-        Assert.assertTrue(viewHolder.checkTrailingWhitespace(test2));
-        Assert.assertEquals(expected2, viewHolder.removeTrailingWhitespace(test2))
+        Assert.assertEquals(expected2, viewHolder.removeLeadingAndTrailingWhitespace(test2))
 
         // No whitespace
         val test3 = "No trailing space";
         val expected3 = "No trailing space";
-        Assert.assertFalse(viewHolder.checkTrailingWhitespace(test3))
-        Assert.assertEquals(expected3, viewHolder.removeTrailingWhitespace(test3))
+        Assert.assertEquals(expected3, viewHolder.removeLeadingAndTrailingWhitespace(test3))
+
+        // blank string
+        val test4 = " \r \t  "
+        val expected4 = "";
+        Assert.assertEquals(expected4, viewHolder.removeLeadingAndTrailingWhitespace(test4))
     }
 
     @Test
-    fun testRemoveTrailingInstanceTab() {
-        val test = "test\t"
+    fun testRemoveLeadingAndTrailingInstanceTab() {
+        val test = "\ttest\t"
         val expected = "test"
-        Assert.assertTrue(viewHolder.checkTrailingWhitespace(test));
-        Assert.assertEquals(expected, viewHolder.removeTrailingWhitespace(test))
+        Assert.assertEquals(expected, viewHolder.removeLeadingAndTrailingWhitespace(test))
     }
 
     @Test
-    fun testRemoveTrailingCarriageReturn() {
-        val test = "test\r"
+    fun testRemoveLeadingAndTrailingCarriageReturn() {
+        val test = "\rtest\r"
         val expected = "test"
-        Assert.assertTrue(viewHolder.checkTrailingWhitespace(test));
-        Assert.assertEquals(expected, viewHolder.removeTrailingWhitespace(test))
+        Assert.assertEquals(expected, viewHolder.removeLeadingAndTrailingWhitespace(test))
     }
 
     @Test
@@ -289,9 +289,8 @@ class UploadMediaDetailAdapterUnitTest {
         val expected1 = "テスト テスト"
         Assert.assertEquals(expected1, viewHolder.convertIdeographicSpaceToLatinSpace(test1));
 
-        val test2 = "テスト　\r　\t　"
+        val test2 = "　\r　\t　テスト　\r　\t　"
         val expected2 = "テスト"
-        Assert.assertTrue(viewHolder.checkTrailingWhitespace(test2));
-        Assert.assertEquals(expected2, viewHolder.removeTrailingWhitespace(test2))
+        Assert.assertEquals(expected2, viewHolder.removeLeadingAndTrailingWhitespace(test2))
     }
 }

